### DISCUSSION TITLE
`C-h .` does nothing

### DISCUSSION THREAD

**Discussion by Inkbottle007** [joaotavora/eglot#954] [Q&A] [answered]
* Server used:               ocaml-lsp
* Emacs version:             29.05.50
* Operating system:          linux
* Eglot version:             commit 2c34dae387c583aa307615779ff570a84d5e723c
* Eglot installation method:       straight-use-package
* Using Doom:                No

#### LSP transcript - M-x eglot-events-buffer (mandatory unless Emacs inoperable)

     
```lisp
[client-request] (id:275) Thu May 12 18:31:00 2022:
(:jsonrpc "2.0" :id 275 :method "textDocument/documentHighlight" :params
          (:textDocument
           (:uri "file:///home/chris/hello.ml")
           :position
           (:line 27 :character 0)))
[server-reply] (id:273) Thu May 12 18:31:00 2022:
(:id 273 :jsonrpc "2.0" :result
     (:signatures
      [(:label "eval : int expr -> int" :parameters
               [(:label
                 [7 15])])]
      :activeSignature 0 :activeParameter 0))
[server-reply] (id:274) Thu May 12 18:31:00 2022:
(:id 274 :jsonrpc "2.0" :result nil)
[server-reply] (id:275) Thu May 12 18:31:00 2022:
(:id 275 :jsonrpc "2.0" :result
     [])
```
    
#### Minimum Reproducible Example (mandatory)
     
`C-h .` does nothing: minibuffer is showing
"C-h (Type ? for further options)."
As you can see in the string above, the "dot" is already included. However it does nothing.

When doing `M-x display-local-help` it says that it is associated with shortcut `C-h .`

When doing `M-x describe-key` `C-h .` it says it is associated with `(eldoc-doc-buffer)`.

On my emacs `eldoc-doc-buffer` and `display-local-help` have totally different actions.

`eldoc-doc-buffer` is associated with action described here: [There should be a cmd called eldoc-doc-buffer that pops up a new buffer showing the contents that would normally be in eldoc. You can scroll in this buffer as well.](https://www.reddit.com/r/emacs/comments/pipu20/comment/hbru60k/?utm_source=share&utm_medium=web2x&context=3)

And here:
```example
eldoc-doc-buffer is an interactive compiled Lisp function in ‘eldoc.el’.

It is bound to C-h ., <f1> . and <help> ..

(eldoc-doc-buffer)

Display ElDoc documentation buffer.

This holds the results of the last documentation request.

  Probably introduced at or before Emacs version 28.1.

[back]

```

`display-local-help` is associated with action: [Eglot relays the diagnostics information received from the server to flymake. Command display-local-help (bound to C-h .) shows the diagnostic message under point, but flymake provides other convenient ways to handle diagnostic errors.](https://github.com/joaotavora/eglot/blob/master/README.md#diagnostics)

Also:
```example
display-local-help is an autoloaded interactive compiled Lisp function in
‘help-at-pt.el’.

Its keys are remapped to ‘eldoc-doc-buffer’.  Without this remapping, it
would be bound to C-h ., <f1> . and <help> ..

(display-local-help &optional ARG)

Display local help in the echo area.
This displays a short help message, namely the string produced by
the ‘kbd-help’ property at point.  If ‘kbd-help’ does not produce
a string, but the ‘help-echo’ property does, then that string is
printed instead.

The string is passed through ‘substitute-command-keys’ before it
is displayed.

A numeric argument ARG prevents display of a message in case
there is no help.  While ARG can be used interactively, it is
mainly meant for use from Lisp.

  Probably introduced at or before Emacs version 22.1.

[back]
```

The keybinding for `display-local-help` is associated to action `eldoc-doc-buffer` in `eglot.el`:
```lisp
(defvar eglot-mode-map
  (let ((map (make-sparse-keymap)))
    (define-key map [remap display-local-help] #'eldoc-doc-buffer)
    map))
```

I think it's a typo, those two actions are not the same, and are not meant to be the same.

`M-x describe-function` show they are associated to the same key binding.

**Comment by joaotavora** [comment]
@Inkbottle007 please DO NOT create any more such Eglot "bug report" issues unless you understand exactly what kind of recipe is requested there.  This is the final time I'm asking you this. It's fine to discuss here.  

I don't understand the problem as it is described (which does not mean there isn't one, just that your description is unclear). In Eglot's default setup `C-h .` is bound to `eldoc-doc-buffer`, not `display-local-help.

**Reply by Inkbottle007** [reply]
https://github.com/joaotavora/eglot/blob/ba618d2cee55c8c339d46621b7e721957cc30a72/README.md#diagnostics
"Eglot relays the diagnostics information received from the server to `flymake`. Command `display-local-help` (bound to `C-h .`)"

**Reply by nemethf** [reply]
I think that was true, but the readme wasn't updated when Eglot was changed to rebound `C-h .`

**Reply by joaotavora** [reply]
> I think that was true, but the readme wasn't updated when Eglot was changed to rebound C-h .

Just updated the README

**Reply by joaotavora** [reply]
`M-x display-local-help` and `M-x eldoc` are two competing commands in Emacs (not Eglot), which aim to provide the same functionality. They should be merged into one.  This is an Emacs problem, not Eglot's.  

Since ElDoc is much more powerful than `display-local-help`, Eglot  "steals" its keybinding and gives it to ElDoc.  But only in Eglot-managed buffers, not globally.  

This, by the way, is the only binding that Eglot controls by default (which  explains how `eglot-mode-map` fits in 3 lines).  Of course, this being Emacs, you can and should add/remove all manner of bindings to that map.

**Reply by Inkbottle007** [reply]
Alright, I get your logic, from the new Readme.md

However `eldoc` doesn't show diagnostic along with function signature: it only shows function signature.

To access diagnostic at point, I can do `display-local-help`. I can also do `flymake-goto-next-error` but it's not at point. I can do `(flymake-diagnostics (point))`...

**Reply by joaotavora** [reply]
That depends on the `eldoc-documentation-strategy` EDIT and `eldoc-documentation-functions`.  Eglot sets it to "enthusiast" which means it will show the most important piece of documentation at point.  EDIT: Eglot's default order:
```
(eglot--setq-saving eldoc-documentation-functions
                        '(eglot-signature-eldoc-function
                          eglot-hover-eldoc-function))
```
Some people are adding `flymake-eldoc-function` to that list manually and changing the strategy to `eldoc-d-compose`  There is some issue around.

Maybe some of those modifications would make sense by default, but I don't have time to look into that right now.

**Reply by joaotavora** [reply]
IOW, Maybe this patch could make sense:
```diff
diff --git a/eglot.el b/eglot.el
index e8f060c..01295fd 100644
--- a/eglot.el
+++ b/eglot.el
@@ -1637,7 +1637,8 @@ Use `eglot-managed-p' to determine if current buffer is managed.")
     (add-hook 'post-self-insert-hook 'eglot--post-self-insert-hook nil t)
     (add-hook 'pre-command-hook 'eglot--pre-command-hook nil t)
     (eglot--setq-saving eldoc-documentation-functions
-                        '(eglot-signature-eldoc-function
+                        '(flymake-eldoc-function
+                          eglot-signature-eldoc-function
                           eglot-hover-eldoc-function))
     (eglot--setq-saving eldoc-documentation-strategy
                         #'eldoc-documentation-enthusiast)
```

**Reply by joaotavora** [reply]
Then again, I'm reasonably sure this would break someone's workflow (mine at least, but I know how to change it/can adapt).

**Reply by Inkbottle007** [reply]
I've tested both suggestions and they are both working. Changing `eldoc-documentation-strategy` to `eldoc-documentation-compose` is probably the best solution. I've inserted that in my `init.el`:
```lisp
(setq eglot-stay-out-of '(eldoc-documentation-strategy)
      eldoc-documentation-strategy 'eldoc-documentation-compose)
```

The `flymake-eldoc-function` solution is hiding the `eglot-signature-eldoc-function` when applicable.

Googling "eldoc-documentation-strategy" is returning only 19 results, none of which are relevant. GitHub doesn't allow quoted search. A comment about `eldoc-documentation-strategy` in eglot documentation could help.

So it's fine for me. Related to: #889 #641 #648 #659

**Comment by joaotavora** [comment, marked as answer]
@Inkbottle007 summarizes it here:

> I've tested both suggestions and they are both working. Changing eldoc-documentation-strategy to eldoc-documentation-compose is probably the best solution. I've inserted that in my init.el:

(setq eglot-stay-out-of '(eldoc-documentation-strategy)
      eldoc-documentation-strategy 'eldoc-documentation-compose)
